### PR TITLE
Add a debugging v2 pool provider wrapper to quick verify caching v2 pool provider quote latencies improvements

### PR DESCRIPTION
**Pull request by jsy1218**
## What?
One aspect of quote latencies investigation comes down to no [V2PoolProvider caching](https://github.com/Uniswap/routing-api/blob/main/lib/handlers/injector-sor.ts#L272) at all in routing-api, so we are thinking a way to quickly verify this is indeed one bottleneck of the E2E quote latencies, without committing to weeks of engineering effort yet.

## Why?
We have some doubts in `V2PoolsLoad` being the bottleneck of the E2E quote latencies. Hence we want to quickly verify.

## How?
We use a hacky debugRouting config that passes down to smart-order-router https://github.com/Uniswap/smart-order-router/pull/325

## Testing?
Tested in local routing-api, and see debugRouting being pass into `AlphaRouterConfig`. I just need to pass in `debugRouting` as part of of the `debugRoutingConfig` into the query string param:

curl --request GET '<RoutingAPIStack_URL>?tokenInAddress=0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2&tokenInChainId=1&tokenOutAddress=0x1f9840a85d5af5bf1d1762f925bdaddc4201f984&tokenOutChainId=1&amount=100&type=exactIn&unicornSecret=0x0&debugRoutingConfig=%7B%22v2PoolSelection.topN%22%3A0%2C%22v2PoolSelection.topNDirectSwaps%22%3A2%2C%22v2PoolSelection.topNTokenInOut%22%3A2%2C%22v2PoolSelection.topNSecondHop%22%3A0%2C%22v2PoolSelection.topNWithEachBaseToken%22%3A0%2C%22v2PoolSelection.topNWithBaseToken%22%3A0%2C%22v3PoolSelection.topN%22%3A0%2C%22v3PoolSelection.topNDirectSwaps%22%3A2%2C%22v3PoolSelection.topNTokenInOut%22%3A2%2C%22v3PoolSelection.topNSecondHop%22%3A0%2C%22v3PoolSelection.topNWithEachBaseToken%22%3A0%2C%22v3PoolSelection.topNWithBaseToken%22%3A0%2C%22maxSwapsPerPath%22%3A2%2C%22maxSplits%22%3A2%2C%22distributionPercent%22%3A20%2C%22writeToCachedRoutes%22%3Afalse%2C%22debugRouting%22%3Atrue%7D'

## Screenshots (optional)
`debugRouting` passed in from the query string param:
Screenshot 2023-08-20 at 10 50 43 PM
Added logging locally to verify:
Screenshot 2023-08-20 at 10 50 54 PM

## Anything Else?

- We may revert this change after we verified.
- This PR relies on https://github.com/Uniswap/smart-order-router/pull/325 getting published. I tested via [local tarball](https://github.com/Uniswap/routing-api/compare/main...jsy1218/explore-sor-local-tarball) method, so that's how I verified it.
- Hold off on check-in, until we have fixed the kms key for the `unicornKey`. See https://uniswapteam.slack.com/archives/C02D07EM805/p1692586467843939?thread_ts=1692575985.012289&cid=C02D07EM805

### CODE REVIEW

File: lib/handlers/injector-sor.ts [hunk 1]
```diff
@@ -269,7 +270,15 @@ export abstract class InjectorSOR<Router, QueryParams> extends Injector<
           sourceOfTruthPoolProvider: noCacheV3PoolProvider,
         })
 
-        const v2PoolProvider = new V2PoolProvider(chainId, multicall2Provider)
+        const nonCachingv2PoolProvider = new V2PoolProvider(chainId, multicall2Provider)
+        const cachingV2PoolProvider = new CachingV2PoolProvider(
+            chainId,
+            new V2PoolProvider(chainId, multicall2Provider),
```
Comment (mikeki) [inline lib/handlers/injector-sor.ts:276]: Can we reference the no caching object here?

Comment (jsy1218) [reply lib/handlers/injector-sor.ts:276]: aw true.

File: lib/handlers/pools/pool-caching/v2/v2-debug-caching-pool-provider.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  }
+
+  getPoolAddress(tokenA: Token, tokenB: Token): { poolAddress: string; token0: Token; token1: Token } {
+    // this is in-memory operation, can just use the existing non-caching pool provider
+    return this.poolProvider.getPoolAddress(tokenA, tokenB)
+  }
+
+  getPools(tokenPairs: [Token, Token][], providerConfig?: ProviderConfig): Promise<V2PoolAccessor> {
+    if (providerConfig?.debugRouting) {
+      return this.cachingPoolProvider.getPools(tokenPairs, providerConfig)
```
Comment (mikeki) [inline lib/handlers/pools/pool-caching/v2/v2-debug-caching-pool-provider.ts:21]: How do we plan on testing this?

**Review by mikeki — APPROVED** [review]

File: lib/handlers/pools/pool-caching/v2/v2-debug-caching-pool-provider.ts (hunk 2 above)
Comment (jsy1218) [reply lib/handlers/pools/pool-caching/v2/v2-debug-caching-pool-provider.ts:21]: I plan to test this via debug query string param:

curl --request GET '<URL>?tokenInAddress=0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2&tokenInChainId=1&tokenOutAddress=0x1f9840a85d5af5bf1d1762f925bdaddc4201f984&tokenOutChainId=1&amount=100&type=exactIn&unicornSecret=0x0&debugRoutingConfig={"v2PoolSelection.topN":0,"v2PoolSelection.topNDirectSwaps":2,"v2PoolSelection.topNTokenInOut":2,"v2PoolSelection.topNSecondHop":0,"v2PoolSelection.topNWithEachBaseToken":0,"v2PoolSelection.topNWithBaseToken":0,"v3PoolSelection.topN":0,"v3PoolSelection.topNDirectSwaps":2,"v3PoolSelection.topNTokenInOut":2,"v3PoolSelection.topNSecondHop":0,"v3PoolSelection.topNWithEachBaseToken":0,"v3PoolSelection.topNWithBaseToken":0,"maxSwapsPerPath":2,"maxSplits":2,"distributionPercent":20,"writeToCachedRoutes":false,"debugRouting":true}'

and from my perf testing script. My perf testing script already account for the latencies difference, so it's easy to compare apple-to-apple if this gets shipped into prod, and we can collect to perf data from prod traffic volume.

**Pull request merged**